changefeed stucks "redo manager fails to close writer"

### What did you do?

1. run test case: redo_apply_multiple_times, which do following steps:
- create redo log changefeed
- run workload "/go-tpc tpcc --db workload --warehouses 200 -T 50 --host upstream-tidb.cdc-redo-long-duration-tps-1806299-1-173 --port 4000 --parts 1 prepare"
- wait changefeed advnance




### What did you expect to see?

1. changefeed should advance and no panic

### What did you see instead?

1. changefeed stucks, "redo manager fails to close writer" for 1.5 hours, and then "The changefeed won't be restarted as it has been experiencing failures for an extended duration"] [maxElapsedTime=1h30m0s]"

```
[2023/06/10 14:44:28.995 +00:00] [ERROR] [feed_state_manager.go:403] ["processor reports an error"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g] [captureID=ec0c65e2-c365-4436-a863-c551613ef791] [error="{\"time\":\"2023-06-10T14:44:28.994331503Z\",\"addr\":\"upstream-ticdc-0.upstream-ticdc-peer.cdc-redo-long-duration-tps-1806299-1-173.svc:8301\",\"code\":\"CDC:ErrInvalidCheckpointTs\",\"message\":\"[CDC:ErrInvalidCheckpointTs]checkpointTs(442081774061486082) should not larger than resolvedTs(442081769683681281)\"}"]
[2023/06/10 14:44:28.995 +00:00] [INFO] [ddl_puller.go:667] ["close the ddl puller"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g_ddl_puller]
[2023/06/10 14:44:28.995 +00:00] [INFO] [meta_manager.go:423] ["redo metaManager bgFlushMeta exits"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g] [error="context canceled"] [errorVerbose="context canceled\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/redo.(*metaManager).bgFlushMeta\n\tgithub.com/pingcap/tiflow/cdc/redo/meta_manager.go:433\ngithub.com/pingcap/tiflow/cdc/redo.(*metaManager).Run.func1\n\tgithub.com/pingcap/tiflow/cdc/redo/meta_manager.go:158\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1598"]
[2023/06/10 14:44:28.995 +00:00] [WARN] [encoding_worker.go:128] ["redo fileWorkerGroup closed with error"] [error="context canceled"] [errorVerbose="context canceled\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/redo/writer/memory.(*encodingWorkerGroup).runWorker\n\tgithub.com/pingcap/tiflow/cdc/redo/writer/memory/encoding_worker.go:153\ngithub.com/pingcap/tiflow/cdc/redo/writer/memory.(*encodingWorkerGroup).Run.func2\n\tgithub.com/pingcap/tiflow/cdc/redo/writer/memory/encoding_worker.go:135\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1598"]
[2023/06/10 14:44:28.995 +00:00] [WARN] [file_worker.go:142] ["redo file workers closed with error"] [error="context canceled"] [errorVerbose="context canceled\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/redo/writer/memory.(*fileWorkerGroup).bgWriteLogs\n\tgithub.com/pingcap/tiflow/cdc/redo/writer/memory/file_worker.go:193\ngithub.com/pingcap/tiflow/cdc/redo/writer/memory.(*fileWorkerGroup).Run.func2\n\tgithub.com/pingcap/tiflow/cdc/redo/writer/memory/file_worker.go:148\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1598"]
[2023/06/10 14:44:28.995 +00:00] [ERROR] [manager.go:522] ["redo manager fails to close writer"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g] [error="context canceled"] [errorVerbose="context canceled\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/redo/writer/memory.(*encodingWorkerGroup).runWorker\n\tgithub.com/pingcap/tiflow/cdc/redo/writer/memory/encoding_worker.go:153\ngithub.com/pingcap/tiflow/cdc/redo/writer/memory.(*encodingWorkerGroup).Run.func2\n\tgithub.com/pingcap/tiflow/cdc/redo/writer/memory/encoding_worker.go:135\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1598"]

...

[2023/06/10 15:28:34.950 +00:00] [ERROR] [manager.go:522] ["redo manager fails to close writer"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g] [error="context canceled"] [errorVerbose="context canceled\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/redo/writer/memory.(*fileWorkerGroup).bgFlushFileCache\n\tgithub.com/pingcap/tiflow/cdc/redo/writer/memory/file_worker.go:170\ngithub.com/pingcap/tiflow/cdc/redo/writer/memory.(*fileWorkerGroup).Run.func3\n\tgithub.com/pingcap/tiflow/cdc/redo/writer/memory/file_worker.go:152\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1598"]
[2023/06/10 15:28:34.950 +00:00] [INFO] [manager.go:527] ["redo manager closed"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g]
[2023/06/10 15:28:34.950 +00:00] [INFO] [processor.go:1037] ["processor sub-component is in stopping"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g] [name=MounterGroup]
[2023/06/10 15:28:34.950 +00:00] [INFO] [processor.go:1037] ["processor sub-component is in stopping"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g] [name=ddlHandler]
[2023/06/10 15:28:34.996 +00:00] [INFO] [client.go:987] ["stream to store closed"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g] [addr=upstream-tikv-1.upstream-tikv-peer.cdc-redo-long-duration-tps-1806299-1-173.svc:20160] [storeID=12]
[2023/06/10 15:28:34.997 +00:00] [INFO] [client.go:987] ["stream to store closed"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g] [addr=upstream-tikv-1.upstream-tikv-peer.cdc-redo-long-duration-tps-1806299-1-173.svc:20160] [storeID=12]
[2023/06/10 15:28:35.001 +00:00] [INFO] [client.go:987] ["stream to store closed"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g] [addr=upstream-tikv-1.upstream-tikv-peer.cdc-redo-long-duration-tps-1806299-1-173.svc:20160] [storeID=12]
[2023/06/10 15:28:35.001 +00:00] [INFO] [processor.go:927] ["Processor drop sort engine successfully"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g]
[2023/06/10 15:28:35.001 +00:00] [INFO] [processor.go:933] ["Processor try to close agent"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g]
[2023/06/10 15:28:35.001 +00:00] [WARN] [server.go:478] ["topic handler returned error"] [error="[CDC:ErrWorkerPoolHandleCancelled]workerpool handle is cancelled"] [errorVerbose="[CDC:ErrWorkerPoolHandleCancelled]workerpool handle is cancelled\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/workerpool.(*defaultEventHandle).GracefulUnregister.func1\n\tgithub.com/pingcap/tiflow/pkg/workerpool/pool_impl.go:230\ngithub.com/pingcap/tiflow/pkg/workerpool.(*defaultEventHandle).GracefulUnregister\n\tgithub.com/pingcap/tiflow/pkg/workerpool/pool_impl.go:252\ngithub.com/pingcap/tiflow/pkg/p2p.(*MessageServer).run.func2\n\tgithub.com/pingcap/tiflow/pkg/p2p/server.go:262\nruntime.goexit\n\truntime/asm_amd64.s:1598"]
[2023/06/10 15:28:35.001 +00:00] [INFO] [processor.go:939] ["Processor closed agent successfully"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g]
[2023/06/10 15:28:35.001 +00:00] [INFO] [processor.go:949] ["processor closed"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g]
[2023/06/10 15:28:35.001 +00:00] [WARN] [server.go:284] ["handler not found"] [topic=changefeed/default/redo-apply-multiple-ci28n65tojtur782h43g/agent]
[2023/06/10 15:30:16.873 +00:00] [INFO] [grpc_pool_impl.go:244] ["recycle connections in grpc pool"] [address=upstream-tikv-2.upstream-tikv-peer.cdc-redo-long-duration-tps-1806299-1-173.svc:20160]
[2023/06/10 15:30:16.873 +00:00] [INFO] [grpc_pool_impl.go:244] ["recycle connections in grpc pool"] [address=upstream-tikv-0.upstream-tikv-peer.cdc-redo-long-duration-tps-1806299-1-173.svc:20160]
[2023/06/10 15:30:16.874 +00:00] [INFO] [grpc_pool_impl.go:244] ["recycle connections in grpc pool"] [address=upstream-tikv-1.upstream-tikv-peer.cdc-redo-long-duration-tps-1806299-1-173.svc:20160]
[2023/06/10 15:58:26.094 +00:00] [WARN] [feed_state_manager.go:547] ["The changefeed won't be restarted as it has been experiencing failures for an extended duration"] [maxElapsedTime=1h30m0s]
```

2. lots of  error "sinkManager: resolved ts should not less than checkpoint ts" seen
```
[2023/06/10 14:44:40.778 +00:00] [ERROR] [manager.go:941] ["sinkManager: resolved ts should not less than checkpoint ts"] [namespace=default] [changefeed=redo-apply-multiple-ci28n65tojtur782h43g] [span={table_id:91,start_key:7480000000000000ff5b5f720000000000fa,end_key:7480000000000000ff5b5f730000000000fa}] [resolvedTs=442081774048378883] [checkpointTs="{\"Mode\":0,\"Ts\":442081774152974342,\"BatchID\":18446744073709551615}"]
```

[Uploading ticdc_1_1806299.log…]()


### Versions of the cluster

cdc version:
["Welcome to Change Data Capture (CDC)"] [release-version=v7.2.0-alpha] [git-hash=c6bebdeb756344564aa140ada7f8db9206e9745a] [git-branch=heads/refs/tags/v7.2.0-alpha] [utc-build-time="2023-06-10 11:33:58"] [go-version="go version go1.20.3 linux/amd64"] [failpoint-build=false]

--- a/cdc/processor/processor.go
+++ b/cdc/processor/processor.go
@@ -108,7 +108,7 @@ var _ scheduler.TableExecutor = (*processor)(nil)
 // 2. Prepare phase for 2 phase scheduling, `isPrepare` should be true.
 // 3. Replicating phase for 2 phase scheduling, `isPrepare` should be false
 func (p *processor) AddTableSpan(
-	ctx context.Context, span tablepb.Span, startTs model.Ts, isPrepare bool,
+	ctx context.Context, span tablepb.Span, startTs model.Ts, isPrepare bool, barrier *schedulepb.Barrier,
 ) (bool, error) {
 	if !p.checkReadyForMessages() {
 		return false, nil
@@ -143,6 +143,16 @@ func (p *processor) AddTableSpan(
 			// table is `prepared`, and a `isPrepare = false` request indicate that old table should
 			// be stopped on original capture already, it's safe to start replicating data now.
 			if !isPrepare {
+				if p.redo.r.Enabled() {
+					var redoResolvedTs model.Ts
+					if barrier != nil {
+						redoResolvedTs = barrier.GlobalBarrierTs
+					} else {
+						stats := p.sinkManager.r.GetTableStats(span)
+						redoResolvedTs = stats.BarrierTs
+					}
+					p.redo.r.StartTable(span, redoResolvedTs)
+				}
 				if err := p.sinkManager.r.StartTable(span, startTs); err != nil {
 					return false, errors.Trace(err)
 				}
@@ -222,7 +232,8 @@ func (p *processor) IsAddTableSpanFinished(span tablepb.Span, isPrepare bool) bo
 	var tableResolvedTs, tableCheckpointTs uint64
 	var state tablepb.TableState
 	done := func() bool {
-		state, alreadyExist := p.sinkManager.r.GetTableState(span)
+		var alreadyExist bool
+		state, alreadyExist = p.sinkManager.r.GetTableState(span)
 		if alreadyExist {
 			stats := p.sinkManager.r.GetTableStats(span)
 			tableResolvedTs = stats.ResolvedTs

--- a/cdc/processor/processor_test.go
+++ b/cdc/processor/processor_test.go
@@ -213,7 +213,7 @@ func TestTableExecutorAddingTableIndirectly(t *testing.T) {
 
 	// table-1: `preparing` -> `prepared` -> `replicating`
 	span := spanz.TableIDToComparableSpan(1)
-	ok, err := p.AddTableSpan(ctx, span, 20, true)
+	ok, err := p.AddTableSpan(ctx, span, 20, true, nil)
 	require.NoError(t, err)
 	require.True(t, ok)
 	p.sinkManager.r.UpdateBarrierTs(20, nil)
@@ -252,7 +252,7 @@ func TestTableExecutorAddingTableIndirectly(t *testing.T) {
 	require.True(t, ok)
 	require.Equal(t, tablepb.TableStatePrepared, state)
 
-	ok, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 30, true)
+	ok, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 30, true, nil)
 	require.NoError(t, err)
 	require.True(t, ok)
 	stats = p.sinkManager.r.GetTableStats(span)
@@ -261,7 +261,7 @@ func TestTableExecutorAddingTableIndirectly(t *testing.T) {
 	require.Equal(t, model.Ts(20), stats.BarrierTs)
 
 	// Start to replicate table-1.
-	ok, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 30, false)
+	ok, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 30, false, nil)
 	require.NoError(t, err)
 	require.True(t, ok)
 
@@ -357,10 +357,10 @@ func TestProcessorClose(t *testing.T) {
 	tester.MustApplyPatches()
 
 	// add tables
-	done, err := p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 20, false)
+	done, err := p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 20, false, nil)
 	require.Nil(t, err)
 	require.True(t, done)
-	done, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(2), 30, false)
+	done, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(2), 30, false, nil)
 	require.Nil(t, err)
 	require.True(t, done)
 
@@ -396,10 +396,10 @@ func TestProcessorClose(t *testing.T) {
 	tester.MustApplyPatches()
 
 	// add tables
-	done, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 20, false)
+	done, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 20, false, nil)
 	require.Nil(t, err)
 	require.True(t, done)
-	done, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(2), 30, false)
+	done, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(2), 30, false, nil)
 	require.Nil(t, err)
 	require.True(t, done)
 	err = p.Tick(ctx)
@@ -441,10 +441,10 @@ func TestPositionDeleted(t *testing.T) {
 	tester.MustApplyPatches()
 
 	// add table
-	done, err := p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 30, false)
+	done, err := p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(1), 30, false, nil)
 	require.Nil(t, err)
 	require.True(t, done)
-	done, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(2), 40, false)
+	done, err = p.AddTableSpan(ctx, spanz.TableIDToComparableSpan(2), 40, false, nil)
 	require.Nil(t, err)
 	require.True(t, done)
 
@@ -550,7 +550,7 @@ func TestUpdateBarrierTs(t *testing.T) {
 	tester.MustApplyPatches()
 
 	span := spanz.TableIDToComparableSpan(1)
-	done, err := p.AddTableSpan(ctx, span, 5, false)
+	done, err := p.AddTableSpan(ctx, span, 5, false, nil)
 	require.True(t, done)
 	require.Nil(t, err)
 	err = p.Tick(ctx)

--- a/cdc/processor/sinkmanager/redo_log_advancer_test.go
+++ b/cdc/processor/sinkmanager/redo_log_advancer_test.go
@@ -79,6 +79,12 @@ func (m *mockRedoDMLManager) UpdateResolvedTs(ctx context.Context,
 	return nil
 }
 
+func (m *mockRedoDMLManager) StartTable(span tablepb.Span, resolvedTs uint64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.resolvedTss[span.TableID] = resolvedTs
+}
+
 func (m *mockRedoDMLManager) GetResolvedTs(span tablepb.Span) model.Ts {
 	m.mu.Lock()
 	defer m.mu.Unlock()

--- a/cdc/redo/manager.go
+++ b/cdc/redo/manager.go
@@ -102,6 +102,7 @@ func (m *ddlManager) GetResolvedTs() model.Ts {
 type DMLManager interface {
 	redoManager
 	AddTable(span tablepb.Span, startTs uint64)
+	StartTable(span tablepb.Span, startTs uint64)
 	RemoveTable(span tablepb.Span)
 	UpdateResolvedTs(ctx context.Context, span tablepb.Span, resolvedTs uint64) error
 	GetResolvedTs(span tablepb.Span) model.Ts
@@ -303,6 +304,12 @@ func (m *logManager) emitRedoEvents(
 		}
 		return nil
 	})
+}
+
+// StartTable starts a table, which means the table is ready to emit redo events.
+// Note that this function should only be called once when adding a new table to processor.
+func (m *logManager) StartTable(span tablepb.Span, resolvedTs uint64) {
+	m.onResolvedTsMsg(span, resolvedTs)
 }
 
 // UpdateResolvedTs asynchronously updates resolved ts of a single table.

--- a/cdc/scheduler/internal/table_executor.go
+++ b/cdc/scheduler/internal/table_executor.go
@@ -18,6 +18,7 @@ import (
 
 	"github.com/pingcap/tiflow/cdc/model"
 	"github.com/pingcap/tiflow/cdc/processor/tablepb"
+	"github.com/pingcap/tiflow/cdc/scheduler/schedulepb"
 )
 
 // TableExecutor is an abstraction for "Processor".
@@ -30,7 +31,7 @@ type TableExecutor interface {
 	// if `isPrepare` is true, the 1st phase of the 2 phase scheduling protocol.
 	// if `isPrepare` is false, the 2nd phase.
 	AddTableSpan(
-		ctx context.Context, span tablepb.Span, startTs model.Ts, isPrepare bool,
+		ctx context.Context, span tablepb.Span, startTs model.Ts, isPrepare bool, barrier *schedulepb.Barrier,
 	) (done bool, err error)
 
 	// IsAddTableSpanFinished make sure the requested table span is in the proper status

--- a/cdc/scheduler/internal/v3/agent/agent.go
+++ b/cdc/scheduler/internal/v3/agent/agent.go
@@ -211,7 +211,7 @@ func (a *agent) Tick(ctx context.Context) (*schedulepb.Barrier, error) {
 		return nil, errors.Trace(err)
 	}
 
-	responses, err := a.tableM.poll(ctx)
+	responses, err := a.tableM.poll(ctx, barrier)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}

--- a/cdc/scheduler/internal/v3/agent/agent_test.go
+++ b/cdc/scheduler/internal/v3/agent/agent_test.go
@@ -164,7 +164,7 @@ func TestAgentHandleMessageDispatchTable(t *testing.T) {
 	// remove table not exist
 	ctx := context.Background()
 	a.handleMessageDispatchTableRequest(removeTableRequest, processorEpoch)
-	responses, err := a.tableM.poll(ctx)
+	responses, err := a.tableM.poll(ctx, &schedulepb.Barrier{})
 	require.NoError(t, err)
 	require.Len(t, responses, 0)
 
@@ -183,7 +183,7 @@ func TestAgentHandleMessageDispatchTable(t *testing.T) {
 	mockTableExecutor.On("AddTableSpan", mock.Anything, mock.Anything,
 		mock.Anything, mock.Anything).Return(false, nil)
 	a.handleMessageDispatchTableRequest(addTableRequest, processorEpoch)
-	responses, err = a.tableM.poll(ctx)
+	responses, err = a.tableM.poll(ctx, &schedulepb.Barrier{})
 	require.NoError(t, err)
 	require.Len(t, responses, 1)
 
@@ -203,14 +203,14 @@ func TestAgentHandleMessageDispatchTable(t *testing.T) {
 	mockTableExecutor.On("IsAddTableSpanFinished", mock.Anything,
 		mock.Anything, mock.Anything).Return(false, nil)
 	a.handleMessageDispatchTableRequest(addTableRequest, processorEpoch)
-	_, err = a.tableM.poll(ctx)
+	_, err = a.tableM.poll(ctx, &schedulepb.Barrier{})
 	require.NoError(t, err)
 
 	mockTableExecutor.ExpectedCalls = mockTableExecutor.ExpectedCalls[:1]
 	mockTableExecutor.On("IsAddTableSpanFinished", mock.Anything,
 		mock.Anything, mock.Anything).Return(true, nil)
 	a.handleMessageDispatchTableRequest(addTableRequest, processorEpoch)
-	responses, err = a.tableM.poll(ctx)
+	responses, err = a.tableM.poll(ctx, &schedulepb.Barrier{})
 	require.NoError(t, err)
 	require.Len(t, responses, 1)
 
@@ -232,7 +232,7 @@ func TestAgentHandleMessageDispatchTable(t *testing.T) {
 		mock.Anything, mock.Anything).Return(false, nil)
 
 	a.handleMessageDispatchTableRequest(addTableRequest, processorEpoch)
-	responses, err = a.tableM.poll(ctx)
+	responses, err = a.tableM.poll(ctx, &schedulepb.Barrier{})
 	require.NoError(t, err)
 	require.Len(t, responses, 1)
 
@@ -247,7 +247,7 @@ func TestAgentHandleMessageDispatchTable(t *testing.T) {
 	mockTableExecutor.On("IsAddTableSpanFinished", mock.Anything,
 		mock.Anything, mock.Anything).Return(true, nil)
 	a.handleMessageDispatchTableRequest(addTableRequest, processorEpoch)
-	responses, err = a.tableM.poll(ctx)
+	responses, err = a.tableM.poll(ctx, &schedulepb.Barrier{})
 	require.NoError(t, err)
 	require.Len(t, responses, 1)
 
@@ -262,7 +262,7 @@ func TestAgentHandleMessageDispatchTable(t *testing.T) {
 		Return(false)
 	// remove table in the replicating state failed, should still in replicating.
 	a.handleMessageDispatchTableRequest(removeTableRequest, processorEpoch)
-	responses, err = a.tableM.poll(ctx)
+	responses, err = a.tableM.poll(ctx, &schedulepb.Barrier{})
 	require.NoError(t, err)
 	require.Len(t, responses, 1)
 	removeTableResponse, ok := responses[0].DispatchTableResponse.
@@ -279,7 +279,7 @@ func TestAgentHandleMessageDispatchTable(t *testing.T) {
 		Return(3, false)
 	// remove table in the replicating state failed, should still in replicating.
 	a.handleMessageDispatchTableRequest(removeTableRequest, processorEpoch)
-	responses, err = a.tableM.poll(ctx)
+	responses, err = a.tableM.poll(ctx, &schedulepb.Barrier{})
 	require.NoError(t, err)
 	require.Len(t, responses, 1)
 	removeTableResponse, ok = responses[0].DispatchTableResponse.
@@ -293,7 +293,7 @@ func TestAgentHandleMessageDispatchTable(t *testing.T) {
 		Return(3, true)
 	// remove table in the replicating state success, should in stopped
 	a.handleMessageDispatchTableRequest(removeTableRequest, processorEpoch)
-	responses, err = a.tableM.poll(ctx)
+	responses, err = a.tableM.poll(ctx, &schedulepb.Barrier{})
 	require.NoError(t, err)
 	require.Len(t, responses, 1)
 	removeTableResponse, ok = responses[0].DispatchTableResponse.
@@ -809,7 +809,7 @@ func TestAgentCommitAddTableDuringStopping(t *testing.T) {
 
 	// Prepare add table is still in-progress.
 	mockTableExecutor.
-		On("AddTableSpan", mock.Anything, mock.Anything, mock.Anything, mock.Anything).
+		On("AddTableSpan", mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything).
 		Return(true, nil).Once()
 	mockTableExecutor.
 		On("IsAddTableSpanFinished", mock.Anything, mock.Anything, mock.Anything, mock.Anything).
@@ -819,7 +819,7 @@ func TestAgentCommitAddTableDuringStopping(t *testing.T) {
 	require.Len(t, trans.SendBuffer, 0)
 
 	mockTableExecutor.
-		On("AddTableSpan", mock.Anything, mock.Anything, mock.Anything, mock.Anything).
+		On("AddTableSpan", mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything).
 		Return(true, nil).Once()
 	mockTableExecutor.
 		On("IsAddTableSpanFinished", mock.Anything, mock.Anything, mock.Anything, mock.Anything).
@@ -1069,7 +1069,7 @@ func newMockTableExecutor() *MockTableExecutor {
 
 // AddTableSpan adds a table span to the executor.
 func (e *MockTableExecutor) AddTableSpan(
-	ctx context.Context, span tablepb.Span, startTs model.Ts, isPrepare bool,
+	ctx context.Context, span tablepb.Span, startTs model.Ts, isPrepare bool, _ *schedulepb.Barrier,
 ) (bool, error) {
 	log.Info("AddTableSpan",
 		zap.String("span", span.String()),

--- a/cdc/scheduler/internal/v3/agent/table.go
+++ b/cdc/scheduler/internal/v3/agent/table.go
@@ -170,14 +170,14 @@ func (t *tableSpan) handleRemoveTableTask() *schedulepb.Message {
 }
 
 func (t *tableSpan) handleAddTableTask(
-	ctx context.Context,
+	ctx context.Context, barrier *schedulepb.Barrier,
 ) (result *schedulepb.Message, err error) {
 	state, _ := t.getAndUpdateTableSpanState()
 	changed := true
 	for changed {
 		switch state {
 		case tablepb.TableStateAbsent:
-			done, err := t.executor.AddTableSpan(ctx, t.task.Span, t.task.StartTs, t.task.IsPrepare)
+			done, err := t.executor.AddTableSpan(ctx, t.task.Span, t.task.StartTs, t.task.IsPrepare, barrier)
 			if err != nil || !done {
 				log.Warn("schedulerv3: agent add table failed",
 					zap.String("namespace", t.changefeedID.Namespace),
@@ -208,7 +208,7 @@ func (t *tableSpan) handleAddTableTask(
 			}
 
 			if t.task.status == dispatchTableTaskReceived {
-				done, err := t.executor.AddTableSpan(ctx, t.task.Span, t.task.StartTs, false)
+				done, err := t.executor.AddTableSpan(ctx, t.task.Span, t.task.StartTs, false, barrier)
 				if err != nil || !done {
 					log.Warn("schedulerv3: agent add table failed",
 						zap.String("namespace", t.changefeedID.Namespace),
@@ -283,14 +283,14 @@ func (t *tableSpan) injectDispatchTableTask(task *dispatchTableTask) {
 		zap.Any("ignoredTask", task))
 }
 
-func (t *tableSpan) poll(ctx context.Context) (*schedulepb.Message, error) {
+func (t *tableSpan) poll(ctx context.Context, barrier *schedulepb.Barrier) (*schedulepb.Message, error) {
 	if t.task == nil {
 		return nil, nil
 	}
 	if t.task.IsRemove {
 		return t.handleRemoveTableTask(), nil
 	}
-	return t.handleAddTableTask(ctx)
+	return t.handleAddTableTask(ctx, barrier)
 }
 
 type tableSpanManager struct {
@@ -310,12 +310,12 @@ func newTableSpanManager(
 	}
 }
 
-func (tm *tableSpanManager) poll(ctx context.Context) ([]*schedulepb.Message, error) {
+func (tm *tableSpanManager) poll(ctx context.Context, barrier *schedulepb.Barrier) ([]*schedulepb.Message, error) {
 	result := make([]*schedulepb.Message, 0)
 	var err error
 	toBeDropped := []tablepb.Span{}
 	tm.tables.Ascend(func(span tablepb.Span, table *tableSpan) bool {
-		message, err1 := table.poll(ctx)
+		message, err1 := table.poll(ctx, barrier)
 		if err != nil {
 			err = errors.Trace(err1)
 			return false

